cpp/dolfinx/fem/Form.h

```diff
@@ -39,17 +41,109 @@ enum class IntegralType : std::int8_t
   vertex = 3          ///< Vertex
 };
 
+namespace impl
+{
+/// @brief Compute the list of entity indices in `mesh` for the ith
+/// integral (kernel) of a given type (i.e. cell, exterior facet, or
+/// interior facet).
+inline std::vector<std::int32_t>
+compute_domain(IntegralType type, const mesh::Topology& topology, int codim,
+               std::span<const std::int32_t> entities,
+               std::span<const std::int32_t> entity_map)
+{
+  assert(codim >= 0);
+  const int tdim = topology.dim();
+
+  std::vector<std::int32_t> mapped_entities;
+  mapped_entities.reserve(entities.size());
+  switch (type)
+  {
+  case IntegralType::cell:
+  {
+    std::ranges::transform(entities, std::back_inserter(mapped_entities),
+                           [&entity_map](auto e) { return entity_map[e]; });
```

Comment:
Until we have resolved https://github.com/FEniCS/dolfinx/issues/3371 we need to have a check that `entity_map` does not map to a negative number, as this would cause a segfault later in the code.

Reply:
Why would it be negative?

This code is unchanged from `main`,  it's just just been moved into a different function: https://github.com/FEniCS/dolfinx/blob/ead49d0fdd57d2273162ede3484505fa77981001/cpp/dolfinx/fem/Form.h#L428

Reply:
Entity maps with it current design, stores -1 for entities of the parent mesh that does not exist in the submesh. That is why I have suggested we use a `MeshTags` instance instead, as it allows for a sparse 1-1 map.